### ISSUE TITLE
ENH: Function to generate all permutations of an array

### ISSUE THREAD

**Issue by timhoffm** [numpy/numpy#26206]
### Proposed new feature or change:

I would like to be able to generate all permutations of an array; e.g.

```
>>> np.permutations([0, 1, 2])
array([[0, 1, 2],
       [0, 2, 1],
       [1, 0, 2],
       [1, 2, 0],
       [2, 0, 1],
       [2, 1, 0]])
````
AFAICS, there's neither a dedicated function for that in numpy already nor is it easily expressable through a combination of existing functions. I've only found the `choice` and `permutation` functions from the random generators, but these only create samples not a  complete set of permutations.

Would such a function be of interest for numpy? If so, I could start with a PR proposal.

**Comment by timhoffm** [comment]
> There is maybe also a bit of a question of whether you need to have some option to gracefully deal with a huge number of permutations (since this can quickly explode in size).

This is a good idea since people may not be aware that this becomes a significant issue with increasing N. One option would be a default size limit (in N or memory usage), and if people want bigger arrays, they have to consciously opt in. An advanced feature would be some batch mode where you receive a generator for a sequence of subarrays of the permutations. But to keep things simple that's likely not for the first shot.

**Comment by charris** [comment]
I would be inclined to make all those functions part of a separate package. Knuth has covered the topic in volumes 4A, 4B, 4C, 4D, 4E, and 4F  of TAOCP, of which 4A and 4B have been published :) I think 4A covers what you want.

**Comment by ngoldbaum** [comment]
I agree this definitely makes sense in an external package. Starting with an external implementation that we could think about importing into numpy would also motivate the discussion about including it in numpy.

Note that a cython implementation would probably be less appealing than a C or C++ implementation - cython produces very large code sizes and I don't know if there's a ton of appetite for adding more cython code to numpy outside of the `random` module.

**Comment by timhoffm** [comment]
> I agree this definitely makes sense in an external package

Are we talking about `permutations()` or a broader set of combinatorial functions like [proposed above](https://github.com/numpy/numpy/issues/26206#issuecomment-2039633646)?

I would be willing to contribute a `permutations()` function to numpy, but I don't have motivation or capacity to start a new package.

> Note that a cython implementation would probably be less appealing than a C or C++ implementation

[As mentioned above](https://github.com/numpy/numpy/issues/26206#issuecomment-2040618323), I see the first priority in ease of use and expressiveness. Getting the API and tests right would be the first step for me: e.g. to be discussed whether one wants one or both of permutations of the first N integers (`permutations(N)`) and permutations of a given array (`permutations([2, 4, 6])`). The initial implementation may even be in terms of `itertools.permutation`. Improving performance is a second step. Whether that works in pure numpy reasonably well or needs C/C++/cython is an optimization detail.

**Comment by timhoffm** [comment]
> The conversion of generated tuples into a Python list remains stable, while conversions into numpy arrays deteriorate beyond set size n>10.
> Can you confirm this observed behavior?

I don't follow this argument. The number of elements scales with n!. The change 9 -> 10 is ~10x for all cases and 10 --> 11 is ~11x.

---

Anyway, coming back to the original question: **Is there interest in having a `np.permutations()` function?**
Any performance discussion is second to that and can be clarified as an implementation detail and/or be optimized later.
